TestPullInvalidIgnoredSR flakyness + data race

It's not a race in our code, but calling `t.Fatal` from many goroutines or something which is not something that we're allowed. (And also a flakyness because this doesn't happen all the time either.)

```
[09:00:37]	[Step 4/4] ==================
[09:00:37]	[Step 4/4] WARNING: DATA RACE
[09:00:37]	[Step 4/4] Write at 0x00c42482c051 by goroutine 114:
[09:00:37]	[Step 4/4]   testing.(*common).FailNow()
[09:00:37]	[Step 4/4]       /usr/local/oldgo/src/testing/testing.go:552 +0x4a
[09:00:37]	[Step 4/4]   testing.(*common).Fatalf()
[09:00:37]	[Step 4/4]       /usr/local/oldgo/src/testing/testing.go:597 +0x94
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/lib/model.pullInvalidIgnored()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/lib/model/requests_test.go:414 +0x12e5
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/lib/model.TestPullInvalidIgnoredSR()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/lib/model/requests_test.go:284 +0x41
[09:00:37]	[Step 4/4]   testing.tRunner()
[09:00:37]	[Step 4/4]       /usr/local/oldgo/src/testing/testing.go:777 +0x16d
[09:00:37]	[Step 4/4] 
[09:00:37]	[Step 4/4] Previous write at 0x00c42482c051 by goroutine 254:
[09:00:37]	[Step 4/4]   testing.(*common).FailNow()
[09:00:37]	[Step 4/4]       /usr/local/oldgo/src/testing/testing.go:552 +0x4a
[09:00:37]	[Step 4/4]   testing.(*common).Fatal()
[09:00:37]	[Step 4/4]       /usr/local/oldgo/src/testing/testing.go:591 +0x7c
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/lib/model.pullInvalidIgnored.func2()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/lib/model/requests_test.go:398 +0xa19
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/lib/model.(*fakeConnection).IndexUpdate()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/lib/model/fakeconns_test.go:80 +0x10b
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/lib/model.(*indexSender).sendIndexTo.func1()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/lib/model/model.go:2001 +0x488
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/lib/model.(*fileInfoBatch).flush()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/lib/model/model.go:2726 +0x90
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/lib/model.(*indexSender).sendIndexTo()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/lib/model/model.go:2047 +0x4e5
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/lib/model.(*indexSender).serve()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/lib/model/model.go:1974 +0x7e0
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/lib/model.(*indexSender).(github.com/syncthing/syncthing/lib/model.serve)-fm()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/lib/model/model.go:1183 +0x4b
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/lib/util.AsService.func1()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/lib/util/utils.go:182 +0x46
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/lib/util.(*service).Serve()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/lib/util/utils.go:233 +0x1b8
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/lib/model.(*indexSender).Serve()
[09:00:37]	[Step 4/4]       <autogenerated>:1 +0x57
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).runService.func1()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:600 +0x6b
[09:00:37]	[Step 4/4] 
[09:00:37]	[Step 4/4] Goroutine 114 (running) created at:
[09:00:37]	[Step 4/4]   testing.(*T).Run()
[09:00:37]	[Step 4/4]       /usr/local/oldgo/src/testing/testing.go:824 +0x564
[09:00:37]	[Step 4/4]   testing.runTests.func1()
[09:00:37]	[Step 4/4]       /usr/local/oldgo/src/testing/testing.go:1063 +0xa4
[09:00:37]	[Step 4/4]   testing.tRunner()
[09:00:37]	[Step 4/4]       /usr/local/oldgo/src/testing/testing.go:777 +0x16d
[09:00:37]	[Step 4/4]   testing.runTests()
[09:00:37]	[Step 4/4]       /usr/local/oldgo/src/testing/testing.go:1061 +0x4e1
[09:00:37]	[Step 4/4]   testing.(*M).Run()
[09:00:37]	[Step 4/4]       /usr/local/oldgo/src/testing/testing.go:978 +0x2cd
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/lib/model.TestMain()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/lib/model/model_test.go:77 +0x8f
[09:00:37]	[Step 4/4]   main.main()
[09:00:37]	[Step 4/4]       _testmain.go:276 +0x22a
[09:00:37]	[Step 4/4] 
[09:00:37]	[Step 4/4] Goroutine 254 (finished) created at:
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).runService()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:588 +0x68
[09:00:37]	[Step 4/4]   github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).Serve()
[09:00:37]	[Step 4/4]       /opt/tcagent/syncthing-1-work/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:476 +0xeaf
[09:00:37]	[Step 4/4] ==================
[09:00:37]	[Step 4/4] 09:00:30 INFO: Connection to AIR6LPZ-7K4PTTV-UXQSMUU-CPQ5YWH-OEDFIIQ-JUG777G-2YQXXR5-YD6AWQR at  closed: Syncthing is being stopped
[09:00:37]	[Step 4/4] --- FAIL: TestPullInvalidIgnoredSR (7.54s)
[09:00:37]	[Step 4/4] 	requests_test.go:398: Didn't receive index updates for all existing files, missing map[ignoredNonExisting:{}]
[09:00:37]	[Step 4/4] 	requests_test.go:414: timed out before index was received
[09:00:37]	[Step 4/4] 	testing.go:730: race detected during execution of test
```

--- a/lib/model/requests_test.go
+++ b/lib/model/requests_test.go
@@ -40,7 +40,7 @@ func TestRequestSimple(t *testing.T) {
 	fc.indexFn = func(folder string, fs []protocol.FileInfo) {
 		select {
 		case <-done:
-			t.Fatalf("More than one index update sent")
+			t.Error("More than one index update sent")
 		default:
 		}
 		for _, f := range fs {
@@ -82,7 +82,7 @@ func TestSymlinkTraversalRead(t *testing.T) {
 	fc.indexFn = func(folder string, fs []protocol.FileInfo) {
 		select {
 		case <-done:
-			t.Fatalf("More than one index update sent")
+			t.Error("More than one index update sent")
 		default:
 		}
 		for _, f := range fs {
@@ -188,7 +188,8 @@ func TestRequestCreateTmpSymlink(t *testing.T) {
 				if f.IsInvalid() {
 					goodIdx <- struct{}{}
 				} else {
-					t.Fatal("Received index with non-invalid temporary file")
+					t.Error("Received index with non-invalid temporary file")
+					close(goodIdx)
 				}
 				return
 			}
@@ -372,7 +373,8 @@ func pullInvalidIgnored(t *testing.T, ft config.FolderType) {
 		secondIndex = true
 		for _, f := range fs {
 			if _, ok := expected[f.Name]; !ok {
-				t.Fatalf("Unexpected file %v was updated in index", f.Name)
+				t.Errorf("Unexpected file %v was updated in index", f.Name)
+				continue
 			}
 			if f.IsInvalid() {
 				t.Errorf("File %v is still marked as invalid", f.Name)
@@ -395,7 +397,7 @@ func pullInvalidIgnored(t *testing.T, ft config.FolderType) {
 		if len(expected) == 0 {
 			close(done)
 		} else if secondIndex {
-			t.Fatal("Didn't receive index updates for all existing files, missing", expected)
+			t.Error("Didn't receive index updates for all existing files, missing", expected)
 		}
 	}
 	// Make sure pulling doesn't interfere, as index updates are racy and
@@ -420,18 +422,22 @@ func TestIssue4841(t *testing.T) {
 	m, fc, fcfg := setupModelWithConnection()
 	defer cleanupModelAndRemoveDir(m, fcfg.Filesystem().URI())
 
-	received := make(chan protocol.FileInfo)
+	received := make(chan []protocol.FileInfo)
 	fc.mut.Lock()
-	fc.indexFn = func(folder string, fs []protocol.FileInfo) {
+	fc.indexFn = func(_ string, fs []protocol.FileInfo) {
+		received <- fs
+	}
+	fc.mut.Unlock()
+	checkReceived := func(fs []protocol.FileInfo) protocol.FileInfo {
+		t.Helper()
 		if len(fs) != 1 {
 			t.Fatalf("Sent index with %d files, should be 1", len(fs))
 		}
 		if fs[0].Name != "foo" {
 			t.Fatalf(`Sent index with file %v, should be "foo"`, fs[0].Name)
 		}
-		received <- fs[0]
+		return fs[0]
 	}
-	fc.mut.Unlock()
 
 	// Setup file from remote that was ignored locally
 	folder := m.folderRunners[defaultFolderConfig.ID].(*sendReceiveFolder)
@@ -441,14 +447,16 @@ func TestIssue4841(t *testing.T) {
 		LocalFlags: protocol.FlagLocalIgnored,
 		Version:    protocol.Vector{}.Update(device1.Short()),
 	}})
-	<-received
+
+	checkReceived(<-received)
 
 	// Scan without ignore patterns with "foo" not existing locally
 	if err := m.ScanFolder("default"); err != nil {
 		t.Fatal("Failed scanning:", err)
 	}
 
-	f := <-received
+	f := checkReceived(<-received)
+
 	if expected := (protocol.Vector{}.Update(myID.Short())); !f.Version.Equal(expected) {
 		t.Errorf("Got Version == %v, expected %v", f.Version, expected)
 	}
@@ -463,25 +471,29 @@ func TestRescanIfHaveInvalidContent(t *testing.T) {
 
 	must(t, ioutil.WriteFile(filepath.Join(tmpDir, "foo"), payload, 0777))
 
-	received := make(chan protocol.FileInfo)
+	received := make(chan []protocol.FileInfo)
 	fc.mut.Lock()
-	fc.indexFn = func(folder string, fs []protocol.FileInfo) {
+	fc.indexFn = func(_ string, fs []protocol.FileInfo) {
+		received <- fs
+	}
+	fc.mut.Unlock()
+	checkReceived := func(fs []protocol.FileInfo) protocol.FileInfo {
+		t.Helper()
 		if len(fs) != 1 {
 			t.Fatalf("Sent index with %d files, should be 1", len(fs))
 		}
 		if fs[0].Name != "foo" {
 			t.Fatalf(`Sent index with file %v, should be "foo"`, fs[0].Name)
 		}
-		received <- fs[0]
+		return fs[0]
 	}
-	fc.mut.Unlock()
 
 	// Scan without ignore patterns with "foo" not existing locally
 	if err := m.ScanFolder("default"); err != nil {
 		t.Fatal("Failed scanning:", err)
 	}
 
-	f := <-received
+	f := checkReceived(<-received)
 	if f.Blocks[0].WeakHash != 103547413 {
 		t.Fatalf("unexpected weak hash: %d != 103547413", f.Blocks[0].WeakHash)
 	}
@@ -506,7 +518,8 @@ func TestRescanIfHaveInvalidContent(t *testing.T) {
 	}
 
 	select {
-	case f := <-received:
+	case fs := <-received:
+		f := checkReceived(fs)
 		if f.Blocks[0].WeakHash != 41943361 {
 			t.Fatalf("unexpected weak hash: %d != 41943361", f.Blocks[0].WeakHash)
 		}
@@ -598,14 +611,24 @@ func TestRequestSymlinkWindows(t *testing.T) {
 	m, fc, fcfg := setupModelWithConnection()
 	defer cleanupModelAndRemoveDir(m, fcfg.Filesystem().URI())
 
-	done := make(chan struct{})
+	received := make(chan []protocol.FileInfo)
 	fc.mut.Lock()
 	fc.indexFn = func(folder string, fs []protocol.FileInfo) {
 		select {
-		case <-done:
-			t.Fatalf("More than one index update sent")
+		case <-received:
+			t.Error("More than one index update sent")
 		default:
 		}
+		received <- fs
+	}
+	fc.mut.Unlock()
+
+	fc.addFile("link", 0644, protocol.FileInfoTypeSymlink, nil)
+	fc.sendIndexUpdate()
+
+	select {
+	case fs := <-received:
+		close(received)
 		// expected first index
 		if len(fs) != 1 {
 			t.Fatalf("Expected just one file in index, got %v", fs)
@@ -617,15 +640,6 @@ func TestRequestSymlinkWindows(t *testing.T) {
 		if !f.IsInvalid() {
 			t.Errorf(`File info was not marked as invalid`)
 		}
-		close(done)
-	}
-	fc.mut.Unlock()
-
-	fc.addFile("link", 0644, protocol.FileInfoTypeSymlink, nil)
-	fc.sendIndexUpdate()
-
-	select {
-	case <-done:
 	case <-time.After(time.Second):
 		t.Fatalf("timed out before pull was finished")
 	}
@@ -667,18 +681,15 @@ func TestRequestRemoteRenameChanged(t *testing.T) {
 	tmpDir := tfs.URI()
 	defer cleanupModelAndRemoveDir(m, tfs.URI())
 
-	done := make(chan struct{})
+	received := make(chan []protocol.FileInfo)
 	fc.mut.Lock()
 	fc.indexFn = func(folder string, fs []protocol.FileInfo) {
 		select {
-		case <-done:
-			t.Fatalf("More than one index update sent")
+		case <-received:
+			t.Error("More than one index update sent")
 		default:
 		}
-		if len(fs) != 2 {
-			t.Fatalf("Received index with %v indexes instead of 2", len(fs))
-		}
-		close(done)
+		received <- fs
 	}
 	fc.mut.Unlock()
 
@@ -694,7 +705,11 @@ func TestRequestRemoteRenameChanged(t *testing.T) {
 	}
 	fc.sendIndexUpdate()
 	select {
-	case <-done:
+	case fs := <-received:
+		close(received)
+		if len(fs) != 2 {
+			t.Fatalf("Received index with %v indexes instead of 2", len(fs))
+		}
 	case <-time.After(10 * time.Second):
 		t.Fatal("timed out")
 	}
@@ -704,38 +719,15 @@ func TestRequestRemoteRenameChanged(t *testing.T) {
 	}
 
 	var gotA, gotB, gotConfl bool
-	done = make(chan struct{})
+	received = make(chan []protocol.FileInfo)
 	fc.mut.Lock()
 	fc.indexFn = func(folder string, fs []protocol.FileInfo) {
 		select {
-		case <-done:
-			t.Fatalf("Received more index updates than expected")
+		case <-received:
+			t.Error("More index update sent than expected")
 		default:
 		}
-		for _, f := range fs {
-			switch {
-			case f.Name == a:
-				if gotA {
-					t.Error("Got more than one index update for", f.Name)
-				}
-				gotA = true
-			case f.Name == b:
-				if gotB {
-					t.Error("Got more than one index update for", f.Name)
-				}
-				gotB = true
-			case strings.HasPrefix(f.Name, "b.sync-conflict-"):
-				if gotConfl {
-					t.Error("Got more than one index update for conflicts of", f.Name)
-				}
-				gotConfl = true
-			default:
-				t.Error("Got unexpected file in index update", f.Name)
-			}
-		}
-		if gotA && gotB && gotConfl {
-			close(done)
-		}
+		received <- fs
 	}
 	fc.mut.Unlock()
 
@@ -763,7 +755,31 @@ func TestRequestRemoteRenameChanged(t *testing.T) {
 	fc.mut.Unlock()
 	fc.sendIndexUpdate()
 	select {
-	case <-done:
+	case fs := <-received:
+		for _, f := range fs {
+			switch {
+			case f.Name == a:
+				if gotA {
+					t.Error("Got more than one index update for", f.Name)
+				}
+				gotA = true
+			case f.Name == b:
+				if gotB {
+					t.Error("Got more than one index update for", f.Name)
+				}
+				gotB = true
+			case strings.HasPrefix(f.Name, "b.sync-conflict-"):
+				if gotConfl {
+					t.Error("Got more than one index update for conflicts of", f.Name)
+				}
+				gotConfl = true
+			default:
+				t.Error("Got unexpected file in index update", f.Name)
+			}
+		}
+		if gotA && gotB && gotConfl {
+			close(received)
+		}
 	case <-time.After(10 * time.Second):
 		t.Errorf("timed out without receiving all expected index updates")
 	}
@@ -890,7 +906,7 @@ func TestRequestDeleteChanged(t *testing.T) {
 	fc.indexFn = func(folder string, fs []protocol.FileInfo) {
 		select {
 		case <-done:
-			t.Fatalf("More than one index update sent")
+			t.Error("More than one index update sent")
 		default:
 		}
 		close(done)
@@ -913,7 +929,7 @@ func TestRequestDeleteChanged(t *testing.T) {
 	fc.indexFn = func(folder string, fs []protocol.FileInfo) {
 		select {
 		case <-done:
-			t.Fatalf("More than one index update sent")
+			t.Error("More than one index update sent")
 		default:
 		}
 		close(done)

--- a/lib/model/requests_test.go
+++ b/lib/model/requests_test.go
@@ -719,45 +719,17 @@ func TestRequestRemoteRenameChanged(t *testing.T) {
 	}
 
 	var gotA, gotB, gotConfl bool
-	received = make(chan []protocol.FileInfo)
 	bIntermediateVersion := protocol.Vector{}.Update(fc.id.Short()).Update(myID.Short())
 	bFinalVersion := bIntermediateVersion.Copy().Update(fc.id.Short())
+	done := make(chan struct{})
 	fc.mut.Lock()
 	fc.indexFn = func(folder string, fs []protocol.FileInfo) {
 		select {
-		case <-received:
-			t.Error("More index update sent than expected")
+		case <-done:
+			t.Error("Received more index updates than expected")
+			return
 		default:
 		}
-		received <- fs
-	}
-	fc.mut.Unlock()
-
-	fd, err := tfs.OpenFile(b, fs.OptReadWrite, 0644)
-	if err != nil {
-		t.Fatal(err)
-	}
-	otherData := []byte("otherData")
-	if _, err = fd.Write(otherData); err != nil {
-		t.Fatal(err)
-	}
-	fd.Close()
-
-	// rename
-	fc.deleteFile(a)
-	fc.updateFile(b, 0644, protocol.FileInfoTypeFile, data[a])
-	// Make sure the remote file for b is newer and thus stays global -> local conflict
-	fc.mut.Lock()
-	for i := range fc.files {
-		if fc.files[i].Name == b {
-			fc.files[i].ModifiedS += 100
-			break
-		}
-	}
-	fc.mut.Unlock()
-	fc.sendIndexUpdate()
-	select {
-	case fs := <-received:
 		for _, f := range fs {
 			switch {
 			case f.Name == a:
@@ -789,8 +761,36 @@ func TestRequestRemoteRenameChanged(t *testing.T) {
 			}
 		}
 		if gotA && gotB && gotConfl {
-			close(received)
+			close(done)
 		}
+	}
+	fc.mut.Unlock()
+
+	fd, err := tfs.OpenFile(b, fs.OptReadWrite, 0644)
+	if err != nil {
+		t.Fatal(err)
+	}
+	otherData := []byte("otherData")
+	if _, err = fd.Write(otherData); err != nil {
+		t.Fatal(err)
+	}
+	fd.Close()
+
+	// rename
+	fc.deleteFile(a)
+	fc.updateFile(b, 0644, protocol.FileInfoTypeFile, data[a])
+	// Make sure the remote file for b is newer and thus stays global -> local conflict
+	fc.mut.Lock()
+	for i := range fc.files {
+		if fc.files[i].Name == b {
+			fc.files[i].ModifiedS += 100
+			break
+		}
+	}
+	fc.mut.Unlock()
+	fc.sendIndexUpdate()
+	select {
+	case <-done:
 	case <-time.After(10 * time.Second):
 		t.Errorf("timed out without receiving all expected index updates")
 	}

